README.md
# base-compat-migrate [![Hackage](https://img.shields.io/hackage/v/base-compat-migrate.svg)](https://hackage.haskell.org/package/base-compat-migrate) [![Build Status](https://travis-ci.org/bergmark/base-compat-migrate.svg)](https://travis-ci.org/bergmark/base-compat-migrate)

This library is meant as a temporary migration library when moving a
library to
[base-compat](http://hackage.haskell.org/package/base-compat) or
upgrading it. Replace `base-compat` with
[base-compat-batteries](https://hackage.haskell.org/package/base-compat-batteries)
in the rest of the README if you want to use that library instead.

`base-compat` defines backwards compatible versions of some base
modules; If you for example want to make sure `import Prelude` does
the same thing on all GHC versions you can instead `import
Prelude.Compat`. If a Compat module exists it means it has some
backwards compatibility fix, it also re-exports everything that is
unchanged from the corresponding base module.

base-compat notably does not have Compat modules for every base module
so you often need to depend on both base and base-compat. The standard
recommended work-flow is to build your project with the oldest GHC you
want to support and add Compat imports accordingly.

If you don't have quick access to old GHCs it can be error-prone to
migrate to upgrade base-compat since you need to look through all base
imports to see if you should be using Compat versions instead.

## Workflow

This library offers a different work flow:

1. Make sure you are using a GHC version compatible with this project
   (At the time of writing it is GHC 8.4.1)
1. Replace any `base` and `base-compat` dependencies in your project
   with `base-compat-migrate`.
1. Add `default-extensions: NoImplicitPrelude` to each build component
   (libraries, executables, test suites, benchmarks).
1. Compile your project and fix issues:
  * Add any needed `Prelude.Compat` imports
  * If a `base` module can't be found change the import to the
    `Compat` version instead.
1. Finally, replace the `base-compat-migrate` dependency with `base`
   and `base-compat`.

Whenever you wish to support newer versions of `base` or `base-compat`
you can repeat these steps.

### Common Issues with Solutions

* If your package declares Paths modules it won't compile without
  access to base. You can drop the Paths module until the affected
  component builds, and then add it again along with the base
  dependency.

## Dependencies

Since `base-compat-migrate` re-exports modules, we depend on minor
versions of `base` and `base-compat`. A new minor version of either
may require a new release of this library. This project's cabal file
is generated by compiling this package with `-fexe` with
cabal-install, or `stack build` with stack and then running the
generated executable.


CHANGELOG.md
# Changelog for base-compat-migrate

#### 0.2.0.1

* Re-generate cabal file, whoops.

## 0.2.0.0

* Upgrade to `base == 4.13.0.*` and `base-compat == 0.11.1.*`.

### 0.1.1.0

* Upgrade to `base-compat == 0.10.4.*`

#### 0.1.0.1

* Upgrade to `base == 4.11.1.*`

## 0.1.0.0

* Initial release


Generate.hs
#!/usr/bin/env stack
-- stack script --resolver nightly-2018-04-14

{-# OPTIONS -Wall #-}
{-# LANGUAGE NoMonomorphismRestriction #-}
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where

import Control.Monad (unless)
import Data.String (fromString)
import Data.String.Conversions (cs)
import Distribution.ModuleName (ModuleName, components, fromComponents)
import Distribution.PackageDescription.Parsec (parseGenericPackageDescription, readGenericPackageDescription, runParseResult)
import Distribution.PackageDescription.PrettyPrint (writeGenericPackageDescription)
import Distribution.Types.CondTree (condTreeData)
import Distribution.Types.GenericPackageDescription (GenericPackageDescription, condLibrary)
import Distribution.Types.Library (exposedModules, reexportedModules)
import Distribution.Types.ModuleReexport (ModuleReexport (..))
import Distribution.Types.PackageName (PackageName)
import Distribution.Verbosity (silent)
import Network.HTTP.Simple (getResponseBody, getResponseHeader, getResponseStatusCode, httpLBS)
import System.Directory (createDirectoryIfMissing, doesFileExist)
import qualified Data.ByteString.Lazy.Char8 as L8

getExposedModules :: GenericPackageDescription -> [ModuleName]
getExposedModules pd =
  case condLibrary pd of
    Just x -> exposedModules (condTreeData x)
    Nothing -> error "No library"

downloadFile :: String -> FilePath -> IO [ModuleName]
downloadFile from to = do
  ex <- doesFileExist to
  bs <- if ex
  then do
    putStrLn $ "Re-using existing " ++ to
    L8.readFile to
  else do
    putStrLn $ "Downloading: " ++ from
    response <- httpLBS $ fromString from
    putStrLn $ "Status: " ++ show (getResponseStatusCode response)
    print $ getResponseHeader "Content-Type" response
    let body = getResponseBody response
    L8.writeFile to body
    pure body
  case runParseResult (parseGenericPackageDescription (cs bs)) of
    (_, Right res) -> do
      putStrLn $ "Parsed " ++ to
      pure $ getExposedModules res
    err -> error $ "parse error: " ++ show err

setReexportedModules :: GenericPackageDescription -> [ModuleReexport] -> GenericPackageDescription
setReexportedModules pd mods =
  pd
    { condLibrary =
      case condLibrary pd of
        Just cn -> Just cn { condTreeData = (condTreeData cn) { reexportedModules = mods } }
        Nothing -> Nothing
    }

mkReexport :: PackageName -> ModuleName -> ModuleReexport
mkReexport pkg m = ModuleReexport (Just pkg) m m

cabalFileUrl :: String -> String -> String
cabalFileUrl name version
  =  "http://raw.githubusercontent.com/commercialhaskell/all-cabal-files/hackage/"
  ++ name
  ++ "/" ++ version
  ++ "/" ++ name ++ ".cabal"

migratePath :: FilePath
migratePath = "base-compat-migrate.cabal"

baseVersion :: String
baseVersion = "4.12.0.0"

baseCompatVersion :: String
baseCompatVersion = "0.11.1"

main :: IO ()
main = do
  ex <- doesFileExist migratePath
  unless ex $
    error $ "Couldn't find " ++ migratePath ++ ", are you in the repository root?"
  createDirectoryIfMissing False "tmp"
  allBaseModules                <- downloadFile
    (cabalFileUrl "base" baseVersion)
    ("tmp/base-" ++ baseVersion ++ ".cabal")
  allBaseCompatModules          <- downloadFile
    (cabalFileUrl "base-compat" baseCompatVersion)
    ("tmp/base-compat-" ++ baseCompatVersion ++ ".cabal")
  let compatModules              = filter ((/= "Repl") . last . components) $ allBaseCompatModules
  let compatModulesWithoutCompat = fromComponents . init . components <$> compatModules
  let baseModules                = filter (`notElem` compatModulesWithoutCompat) allBaseModules
  pd <- readGenericPackageDescription silent migratePath
  let pd' = setReexportedModules pd $ map (mkReexport "base") baseModules ++ map (mkReexport "base-compat") compatModules
  writeGenericPackageDescription migratePath pd'


generate-library/Main.hs
#!/usr/bin/env stack
-- stack script --resolver nightly-2018-04-14

{-# OPTIONS -Wall #-}
{-# LANGUAGE NoMonomorphismRestriction #-}
{-# LANGUAGE OverloadedStrings #-}
module Main (main) where

import Control.Monad (unless)
import Data.String (fromString)
import Data.String.Conversions (cs)
import Distribution.ModuleName (ModuleName, components, fromComponents)
import Distribution.PackageDescription.Parsec (parseGenericPackageDescription, readGenericPackageDescription, runParseResult)
import Distribution.PackageDescription.PrettyPrint (writeGenericPackageDescription)
import Distribution.Types.CondTree (condTreeData)
import Distribution.Types.GenericPackageDescription (GenericPackageDescription, condLibrary)
import Distribution.Types.Library (exposedModules, reexportedModules)
import Distribution.Types.ModuleReexport (ModuleReexport (..))
import Distribution.Types.PackageName (PackageName)
import Distribution.Verbosity (silent)
import Network.HTTP.Simple (getResponseBody, getResponseHeader, getResponseStatusCode, httpLBS)
import System.Directory (createDirectoryIfMissing, doesFileExist)
import qualified Data.ByteString.Lazy.Char8 as L8

getExposedModules :: GenericPackageDescription -> [ModuleName]
getExposedModules pd =
  case condLibrary pd of
    Just x -> exposedModules (condTreeData x)
    Nothing -> error "No library"

downloadFile :: String -> FilePath -> IO [ModuleName]
downloadFile from to = do
  ex <- doesFileExist to
  bs <- if ex
  then do
    putStrLn $ "Re-using existing " ++ to
    L8.readFile to
  else do
    putStrLn $ "Downloading: " ++ from
    response <- httpLBS $ fromString from
    putStrLn $ "Status: " ++ show (getResponseStatusCode response)
    print $ getResponseHeader "Content-Type" response
    let body = getResponseBody response
    L8.writeFile to body
    pure body
  case runParseResult (parseGenericPackageDescription (cs bs)) of
    (_, Right res) -> do
      putStrLn $ "Parsed " ++ to
      pure $ getExposedModules res
    err -> error $ "parse error: " ++ show err

setReexportedModules :: GenericPackageDescription -> [ModuleReexport] -> GenericPackageDescription
setReexportedModules pd mods =
  pd
    { condLibrary =
      case condLibrary pd of
        Just cn -> Just cn { condTreeData = (condTreeData cn) { reexportedModules = mods } }
        Nothing -> Nothing
    }

mkReexport :: PackageName -> ModuleName -> ModuleReexport
mkReexport pkg m = ModuleReexport (Just pkg) m m

cabalFileUrl :: String -> String -> String
cabalFileUrl name version
  =  "http://raw.githubusercontent.com/commercialhaskell/all-cabal-files/hackage/"
  ++ name
  ++ "/" ++ version
  ++ "/" ++ name ++ ".cabal"

migratePath :: FilePath
migratePath = "base-compat-migrate.cabal"

baseVersion :: String
baseVersion = "4.11.0.0"

baseCompatVersion :: String
baseCompatVersion = "0.10.1"

main :: IO ()
main = do
  ex <- doesFileExist migratePath
  unless ex $
    error $ "Couldn't find " ++ migratePath ++ ", are you in the repository root?"
  createDirectoryIfMissing False "tmp"
  allBaseModules                <- downloadFile
    (cabalFileUrl "base" baseVersion)
    ("tmp/base-" ++ baseVersion ++ ".cabal")
  allBaseCompatModules          <- downloadFile
    (cabalFileUrl "base-compat" baseCompatVersion)
    ("tmp/base-compat-" ++ baseCompatVersion ++ ".cabal")
  let compatModules              = filter ((== "Compat") . last . components) allBaseCompatModules
  let compatModulesWithoutCompat = fromComponents . init . components <$> compatModules
  let baseModules                = filter (`notElem` compatModulesWithoutCompat) allBaseModules
  pd <- readGenericPackageDescription silent migratePath
  let pd' = setReexportedModules pd $ map (mkReexport "base") baseModules ++ map (mkReexport "base-compat") compatModules
  writeGenericPackageDescription migratePath pd'
